src/main/java/ratpack/zipkin/internal/RatpackCurrentTraceContext.java

```diff
@@ -21,26 +25,47 @@ public RatpackCurrentTraceContext() {
   @Override
   public TraceContext get() {
     return registrySupplier.get()
-        .maybeGet(TraceContextHolder.class)
-        .map(TraceContextHolder::getContext)
+        .maybeGet(ContextHolder.class)
+        .map(h -> h.context)
         .orElse(null);
   }
 
   @Override
   public Scope newScope(TraceContext currentSpan) {
-    final TraceContext previous = get();
-    registrySupplier.get().add(TraceContextHolder.class, new TraceContextHolder(currentSpan));
-    return () -> registrySupplier.get().add(new TraceContextHolder(previous));
+
+    if (currentSpan != null) {
+
+      final Optional<ContextHolder> previous = registrySupplier.get()
+          .maybeGet(ContextHolder.class);
+
+      registrySupplier.get().add(new ContextHolder(currentSpan));
+      MDC.put(MCD_TRACEID_KEY, currentSpan.traceIdString());
+
+      return () -> {
+        if (previous.isPresent() && previous.get().context != null) {
+          registrySupplier.get().add(previous);
+          MDC.put(MCD_TRACEID_KEY, previous.get().context.traceIdString());
+        } else {
+          registrySupplier.get().add(ContextHolder.EMPTY);
+          MDC.remove(MCD_TRACEID_KEY);
+        }
+      };
+    } else {
+      registrySupplier.get().add(ContextHolder.EMPTY);
+      return () -> {};
+    }
   }
 
-  public static final class TraceContextHolder {
+  private static final class ContextHolder {
+
+    private static final ContextHolder EMPTY = new ContextHolder(null);
```

Comment:
An empty value to avoid extra object creation every time the last Scope is closed.